lib/flagsmith/engine/features/models.rb

```diff
@@ -0,0 +1,172 @@
+
+        class << self
+          def build(json)
+            new(
+              **json.slice(:id, :percentage_allocation, :mv_fs_value_uuid)
+                    .merge(multivariate_feature_option: MultivariateOption.build(json[:multivariate_feature_option]))
+            )
+          end
+        end
+      end
+
+      # FeatureSegment
+      class Segment
+        attr_reader :priority
+
+        def initialize(params = {})
+          @priority = params.fetch(:priority, nil).to_i
+        end
+      end
+
+      # FeatureStateModel
+      class State
+        include Flagsmith::Engine::Utils::HashFunc
+
+        attr_reader :feature, :enabled, :django_id, :uuid, :feature_segment
+        attr_accessor :multivariate_feature_state_values
+
+        def initialize(params = {})
+          @feature = params.fetch(:feature)
+          @enabled = params.fetch(:enabled)
+          @feature_segment = params.fetch(:feature_segment, nil)
+          @django_id = params.fetch(:django_id, nil)
+          @feature_state_value = params.fetch(:feature_state_value, nil)
+          @uuid = params.fetch(:uuid, SecureRandom.uuid)
+          @multivariate_feature_state_values = params.fetch(:multivariate_feature_state_values, [])
+        end
+
+        attr_writer :feature_state_value
+
+        def get_value(identity_id = nil)
+          return multivariate_value(identity_id) if identity_id && multivariate_feature_state_values.length.positive?
+
+          @feature_state_value
+        end
+
+        alias set_value feature_state_value=
+        alias feature_state_uuid uuid
+        alias enabled? enabled
+
+        def multivariate_value(identity_id)
+          percentage_value = hashed_percentage_for_object_ids([django_id || uuid, identity_id])
+
+          start_percentage = 0
+          multivariate_feature_state_values.sort.each do |multi_fs_value|
+            limit = multi_fs_value.percentage_allocation + start_percentage
+
+            if start_percentage <= percentage_value && percentage_value < limit
+              return multi_fs_value.multivariate_feature_option.value
+            end
+
+            start_percentage = limit
+          end
+          @feature_state_value
+        end
+
+        # Returns `true` if `self` is higher segment priority than `other`
+        # (i.e. has lower value for feature_segment.priority)
+        # NOTE:
+        #     A segment will be considered higher priority only if:
+        #     1. `other` does not have a feature segment
+        #        (i.e: it is an environment feature state or it's a
+        #        feature state with feature segment but from an old document
+        #        that does not have `feature_segment.priority`)
+        #        but `self` does.
+        #     2. `other` have a feature segment with high priority
+        def higher_segment_priority?(other)
+          feature_segment.priority < other.feature_segment.priority
+        rescue TypeError, NoMethodError
+          false
```

Comment:
I don't know that this quite covers all the use cases does it? 

If `feature_segment` is not null but `other.feature_segment` is, it should return `true` but I think this will return `false` ?